File: src/common/ShoppingList.py
import os
import sys
from uuid import uuid4
from src.common.PNCounter import PNCounter

# Define module
current_path = os.path.dirname(__file__) + '/../..'
sys.path.append(current_path)


class ShoppingList:
    def __init__(self, uuid=None):
        self.items = dict()
        self.uuid = uuid4()
        if uuid is not None:
            self.uuid = uuid

    def inc_or_add_item(self, item_name, quantity, replica):
        item = self.items.get(item_name, PNCounter.zero())
        item.inc(replica, quantity)
        self.items[item_name] = item

    def dec_item(self, item_name, quantity, replica):
        item = self.items.get(item_name, PNCounter.zero())
        item.dec(replica, quantity)
        self.items[item_name] = item

    @staticmethod
    def merge(a, b):

        items = [x for x in set(a.items.keys()) | set(b.items.keys())]

        mergedItems = dict()
        for item in items:
            mergedItems[item] = PNCounter.merge(
                a.items.get(item, PNCounter.zero()),
                b.items.get(item, PNCounter.zero())
            )

        mergedSL = ShoppingList(a.uuid)
        mergedSL.items = mergedItems
        return mergedSL

    def __str__(self):
        items = ", ".join(f"\"{key}\": {{{str(value)}}}" for key, value in self.items.items())
        return f"{{\"uuid\":\"{self.uuid}\", \"items\":{{{items}}}}}".replace("\'", "\"")

File: app/client2/src/js/CRDTs/ShoppingList.js
const { v4: uuidv4 } = require('uuid');
const PNCounter = require("./PNCounter");

class ShoppingList{
    constructor(uuid=null) {
        this.items = new Map()
        this.uuid = uuidv4()
        if(uuid !== null) {
            this.uuid = uuid
        }
    }

    inc_or_add_item(item_name, quantity, replica){
        let item =  this.items.get(item_name) ?? PNCounter.zero()
        item.inc(replica, quantity)
        this.items.set(item_name, item)
    }

    dec_item(item_name, quantity, replica){
        let item =  this.items.get(item_name) ?? PNCounter.zero()
        item.dec(replica, quantity)
        this.items.set(item_name, item)
    }

    toString(){
        let items = Array.from(this.items.entries()).map(([key, value]) => `"${key}": {${value.toString()}}`);
        return "{\"uuid\":\"" + this.uuid + "\", \"items\": {" + items + "}}"
    }

}

module.exports = ShoppingList